storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -95,76 +107,196 @@ func (s *fileSystemSource) shardAvailability(
+						s.log.WithFields(
+							xlog.NewLogField("shard", shard),
+							xlog.NewLogField("error", err),
+						).Error("reading data file failed")
+						hasError = true
+						break
+					}
+
+					encoder := bopts.DatabaseBlockOptions().EncoderPool().Get()
+					encoder.ResetSetData(timeRange.Start, data, false)
+					block := bopts.DatabaseBlockOptions().DatabaseBlockPool().Get()
+					block.Reset(timeRange.Start, encoder)
+					seriesMap.AddBlock(id, block)
+				}
+				if !hasError {
+					if err := r.Validate(); err != nil {
+						s.log.WithFields(
+							xlog.NewLogField("shard", shard),
+							xlog.NewLogField("error", err),
+						).Error("data validation failed")
+						hasError = true
+					}
+				}
+				r.Close()
+				if !hasError {
+					tr = tr.RemoveRange(timeRange)
+				} else {
+					timesWithErrors = append(timesWithErrors, timeRange.Start)
 				}
 			}
-			r.Close()
 
-			if !hasError {
-				seriesMap.AddResult(curMap)
-				tr = tr.RemoveRange(timeRange)
-			} else {
-				curMap.Close()
+			// NB(xichen): this is the exceptional case where we encountered errors due to files
+			// being corrupted, which should be fairly rare so we can live with the overhead. We
+			// experimented with adding the series to a temporary map and only adding the temporary map
+			// to the final result but adding series to large map with string keys is expensive, and
+			// the current implementation saves the extra overhead of merging temporary map with the
+			// final result.
+			if seriesMap != nil && len(timesWithErrors) > 0 {
+				s.log.WithFields(
+					xlog.NewLogField("shard", shard),
+					xlog.NewLogField("timesWithErrors", timesWithErrors),
+				).Info("deleting entries from results for times with errors")
+
+				allSeries := seriesMap.AllSeries()
+				for id, series := range allSeries {
+					for _, t := range timesWithErrors {
+						series.RemoveBlockAt(t)
+					}
+					if series.Len() == 0 {
+						delete(allSeries, id)
+					}
+				}
 			}
-		}
 
-		result.Add(shard, seriesMap, tr)
+			resultLock.Lock()
+			result.Add(shard, seriesMap, tr)
+			resultLock.Unlock()
+		})
+	}
+
+	wg.Wait()
+	return result
+}
+
+func (s *fileSystemSource) Read(
+	namespace string,
+	shardsTimeRanges bootstrap.ShardTimeRanges,
+) (bootstrap.Result, error) {
+	if shardsTimeRanges.IsEmpty() {
+		return nil, nil
 	}
 
+	start := time.Now()
+	s.log.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("numShards", len(shardsTimeRanges)),
+	).Info("started reading data from filesystem for bootstrapping")
+
+	readersCh := make(chan shardReaders)
+
+	go s.enqueueReaders(namespace, shardsTimeRanges, readersCh)
+	result := s.bootstrapFromReaders(readersCh)
+
+	end := time.Now()
+
+	s.log.WithFields(
+		xlog.NewLogField("namespace", namespace),
+		xlog.NewLogField("numShards", len(shardsTimeRanges)),
+		xlog.NewLogField("duration", end.Sub(start).String()),
+	).Info("finished reading data from filesystem for bootstrapping")
```

Comment:
meh, they are somewhat useful but for now i'll move logging into the bootstrap manager